File: src/frontend/platform/src/controllers/API/developerToken.ts
import axios from "@/controllers/request"

export interface DeveloperTokenGlobalConfig {
  ip_whitelist: string
  rate_limit_per_minute: number | null
}

export type DeveloperTokenRouteMatchType = "METHOD_PATH" | "PATH" | "PREFIX"

export interface DeveloperTokenRouteRule {
  match_type: DeveloperTokenRouteMatchType
  method?: string | null
  path: string
}

export type DeveloperTokenFileSyncMode = "fixed" | "dynamic"

export type DeveloperTokenFileSyncDynamicSource =
  | "department_id"
  | "responsible_person_id"

export interface DeveloperTokenFileSyncRule {
  category: {
    code: string
    subcategory_code: string
  }
  business_domain: {
    mode: DeveloperTokenFileSyncMode
    code: string | null
  }
  target_space: {
    mode: DeveloperTokenFileSyncMode
    knowledge_id: number | null
    folder_id: number | null
  }
  dynamic_source: DeveloperTokenFileSyncDynamicSource | null
}

export interface DeveloperTokenFileSyncCategoryOption {
  code: string
  label: string
  children: Array<{ code: string; label: string }>
}

export interface DeveloperTokenFileSyncBusinessDomainOption {
  code: string
  name: string
}

export interface DeveloperTokenFileSyncTargetSpaceOption {
  id: number
  name: string
  selectable: boolean
  has_children: boolean
}

export interface DeveloperTokenFileSyncTargetSpaceGroup {
  space_type: "public" | "department"
  spaces: DeveloperTokenFileSyncTargetSpaceOption[]
}

export interface DeveloperTokenFileSyncCursorPage<T> {
  data: T[]
  has_more: boolean
  next_cursor: string | null
  page_size: number
}

export interface DeveloperTokenFileSyncTargetFolderOption {
  id: number
  name: string
  selectable: boolean
  navigation_only: boolean
  has_children: boolean
}

export interface DeveloperTokenFileSyncTargetDisplay {
  knowledge_id: number
  knowledge_name?: string | null
  target_type: "root" | "folder"
  folder_id?: number | null
  folder_path: Array<{ id: number; name: string }>
  stale: boolean
}

export interface DeveloperTokenFileSyncOptions {
  tenant_id: number
  user_id: number
  categories: DeveloperTokenFileSyncCategoryOption[]
  business_domains: DeveloperTokenFileSyncBusinessDomainOption[]
  target_space_groups: DeveloperTokenFileSyncCursorPage<DeveloperTokenFileSyncTargetSpaceGroup>
}

export interface DeveloperTokenRecord {
  id: number
  tenant_id: number
  tenant_name?: string | null
  user_id: number
  user_name?: string | null
  name: string
  token_prefix: string
  enabled: boolean
  override_ip_whitelist: boolean
  override_rate_limit: boolean
  rate_limit_per_minute?: number | null
  route_rule_count: number
  file_sync_rule?: DeveloperTokenFileSyncRule | null
  file_sync_target_display?: DeveloperTokenFileSyncTargetDisplay | null
  last_used_time?: string | null
  last_used_ip?: string | null
  created_by?: number | null
  updated_by?: number | null
  create_time?: string | null
  update_time?: string | null
}

export interface DeveloperTokenDetail extends DeveloperTokenRecord {
  ip_whitelist?: string | null
  route_whitelist?: DeveloperTokenRouteRule[] | null
}

export interface DeveloperTokenPage {
  data: DeveloperTokenRecord[]
  total: number
}

export interface DeveloperTokenPayload {
  name: string
  user_id?: number
  department_id?: number
  dept_id?: string
  enabled: boolean
  override_ip_whitelist: boolean
  ip_whitelist?: string
  override_rate_limit: boolean
  rate_limit_per_minute?: number | null
  route_whitelist?: DeveloperTokenRouteRule[] | null
  file_sync_rule?: DeveloperTokenFileSyncRule | null
}

export interface DeveloperTokenCreateResponse {
  token: DeveloperTokenRecord
  plaintext_token: string
}

export interface DeveloperTokenSecretResponse {
  id: number
  token_prefix: string
  plaintext_token: string
}

export async function listDeveloperTokensApi(params: {
  page?: number
  limit?: number
  keyword?: string
  tenant_id?: number
  user_id?: number
  enabled?: boolean
} = {}): Promise<DeveloperTokenPage> {
  return await axios.get("/api/v1/admin/developer-tokens", {
    params: { page: 1, limit: 20, ...params },
  })
}

export async function createDeveloperTokenApi(
  data: DeveloperTokenPayload
): Promise<DeveloperTokenCreateResponse> {
  return await axios.post("/api/v1/admin/developer-tokens", data)
}

export async function getDeveloperTokenDetailApi(
  tokenId: number
): Promise<DeveloperTokenDetail> {
  return await axios.get(`/api/v1/admin/developer-tokens/${tokenId}`)
}

export async function updateDeveloperTokenApi(
  tokenId: number,
  data: Partial<DeveloperTokenPayload>
): Promise<DeveloperTokenRecord> {
  return await axios.put(`/api/v1/admin/developer-tokens/${tokenId}`, data)
}

export async function deleteDeveloperTokenApi(tokenId: number): Promise<void> {
  return await axios.delete(`/api/v1/admin/developer-tokens/${tokenId}`)
}

export async function viewDeveloperTokenSecretApi(
  tokenId: number
): Promise<DeveloperTokenSecretResponse> {
  return await axios.get(`/api/v1/admin/developer-tokens/${tokenId}/secret`)
}

export async function getDeveloperTokenGlobalConfigApi(): Promise<DeveloperTokenGlobalConfig> {
  return await axios.get("/api/v1/admin/developer-tokens/config/global")
}

export async function updateDeveloperTokenGlobalConfigApi(
  data: DeveloperTokenGlobalConfig
): Promise<DeveloperTokenGlobalConfig> {
  return await axios.put("/api/v1/admin/developer-tokens/config/global", data)
}

export async function getDeveloperTokenFileSyncOptionsApi(params: {
  tenant_id: number
  user_id: number
  space_cursor?: string
  space_page_size?: number
  space_keyword?: string
  signal?: AbortSignal
}): Promise<DeveloperTokenFileSyncOptions> {
  const { signal, ...query } = params
  return await axios.get("/api/v1/admin/developer-tokens/config/file-sync-options", {
    params: { space_page_size: 50, ...query },
    ...(signal ? { signal } : {}),
  })
}

export async function getDeveloperTokenFileSyncTargetChildrenApi(params: {
  tenant_id: number
  user_id: number
  knowledge_id: number
  parent_id?: number
  cursor?: string
  page_size?: number
  signal?: AbortSignal
}): Promise<DeveloperTokenFileSyncCursorPage<DeveloperTokenFileSyncTargetFolderOption>> {
  const { signal, ...query } = params
  return await axios.get(
    "/api/v1/admin/developer-tokens/config/file-sync-target-children",
    {
      params: { page_size: 50, ...query },
      ...(signal ? { signal } : {}),
    }
  )
}

File: src/backend/bisheng/developer_token/domain/schemas/developer_token.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from bisheng.common.dependencies.user_deps import UserPayload
from bisheng.developer_token.domain.file_sync_folder_path import normalize_file_sync_folder_path
from bisheng.knowledge.domain.constants import normalize_business_domain_code

FileSyncDynamicSource = Literal["department_id", "responsible_person_id"]
FileSyncFolderMode = Literal["none", "fixed", "dynamic"]
FileSyncFolderDynamicSource = Literal["department_name", "caller_main_department_name"]


class FileSyncCategoryRule(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(pattern=r"^[A-Z0-9_]{1,16}$")
    subcategory_code: str = Field(pattern=r"^[A-Z0-9_-]{1,16}$")

    @field_validator("code", "subcategory_code", mode="before")
    @classmethod
    def normalize_code(cls, value: Any) -> Any:
        return value.strip().upper() if isinstance(value, str) else value


class FileSyncBusinessDomainRule(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["fixed", "dynamic"]
    code: str | None = None
    dynamic_source: FileSyncDynamicSource | None = None

    @field_validator("code", mode="before")
    @classmethod
    def normalize_code(cls, value: Any) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("business domain code must be a string")
        if not value.strip():
            return None
        normalized = normalize_business_domain_code(value)
        if normalized is None:
            raise ValueError("business domain code is invalid")
        return normalized


class FileSyncTargetSpaceRule(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["fixed", "dynamic"]
    knowledge_id: int | None = Field(default=None, strict=True, gt=0)
    folder_id: int | None = Field(default=None, strict=True, gt=0)
    dynamic_source: FileSyncDynamicSource | None = None
    folder_mode: FileSyncFolderMode = "none"
    folder_path: str | None = None
    parent_folder_path: str | None = None
    folder_dynamic_source: FileSyncFolderDynamicSource | None = None

    @field_validator("folder_path", "parent_folder_path", mode="before")
    @classmethod
    def normalize_folder_path(cls, value: Any) -> str | None:
        if value is None:
            return None
        if isinstance(value, str) and not value.strip():
            return None
        return normalize_file_sync_folder_path(value)

    @model_validator(mode="after")
    def infer_legacy_folder_mode(self):
        if self.folder_mode == "none":
            if self.folder_path:
                self.folder_mode = "fixed"
            elif self.folder_id is not None:
                self.folder_mode = "fixed"
        return self


class DeveloperTokenFileSyncRule(BaseModel):
    model_config = ConfigDict(extra="forbid")

    category: FileSyncCategoryRule
    business_domain: FileSyncBusinessDomainRule
    target_space: FileSyncTargetSpaceRule
    # Deprecated: migrated into per-dimension dynamic_source on read; never persisted on save.
    dynamic_source: FileSyncDynamicSource | None = None

    @model_validator(mode="before")
    @classmethod
    def migrate_legacy_dynamic_source(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        legacy = data.get("dynamic_source")
        if legacy is None:
            return data
        payload = dict(data)
        business_domain = dict(payload.get("business_domain") or {})
        target_space = dict(payload.get("target_space") or {})
        if business_domain.get("mode") == "dynamic" and business_domain.get("dynamic_source") is None:
            business_domain["dynamic_source"] = legacy
        if target_space.get("mode") == "dynamic" and target_space.get("dynamic_source") is None:
            target_space["dynamic_source"] = legacy
        payload["business_domain"] = business_domain
        payload["target_space"] = target_space
        payload["dynamic_source"] = None
        return payload


class DeveloperTokenPrincipal(BaseModel):
    model_config = ConfigDict(frozen=True, arbitrary_types_allowed=True)

    token_id: int
    token_name: str = ""
    tenant_id: int
    user: UserPayload
    raw_file_sync_rule: dict | None = None


class FileSyncOptionChild(BaseModel):
    code: str
    label: str


class FileSyncOptionCategory(BaseModel):
    code: str
    label: str
    children: list[FileSyncOptionChild]


class FileSyncOptionBusinessDomain(BaseModel):
    code: str
    name: str
    space_ids: list[int] = Field(default_factory=list)


class FileSyncTargetSpaceOption(BaseModel):
    id: int
    name: str
    selectable: bool
    has_children: bool
    business_domain_codes: list[str] = Field(default_factory=list)


class FileSyncTargetSpaceGroup(BaseModel):
    space_type: Literal["public", "department"]
    spaces: list[FileSyncTargetSpaceOption]


class FileSyncTargetSpaceGroupsPage(BaseModel):
    data: list[FileSyncTargetSpaceGroup]
    has_more: bool
    next_cursor: str | None
    page_size: int


class FileSyncTargetFolderOption(BaseModel):
    id: int
    name: str
    selectable: bool
    navigation_only: bool
    has_children: bool


class DeveloperTokenFileSyncTargetChildren(BaseModel):
    data: list[FileSyncTargetFolderOption]
    has_more: bool
    next_cursor: str | None
    page_size: int


class FileSyncTargetPathItem(BaseModel):
    id: int
    name: str


class FileSyncTargetDisplay(BaseModel):
    knowledge_id: int
    knowledge_name: str | None = None
    target_type: Literal["root", "folder"]
    folder_id: int | None = None
    folder_path: list[FileSyncTargetPathItem] = Field(default_factory=list)
    stale: bool = False


class DeveloperTokenFileSyncOptions(BaseModel):
    tenant_id: int
    user_id: int
    categories: list[FileSyncOptionCategory]
    business_domains: list[FileSyncOptionBusinessDomain]
    target_space_groups: FileSyncTargetSpaceGroupsPage


class DeveloperTokenGlobalConfig(BaseModel):
    ip_whitelist: str = ""
    rate_limit_per_minute: int | None = None


class DeveloperTokenListQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=20, ge=1, le=200)
    keyword: str | None = None
    tenant_id: int | None = None
    user_id: int | None = None
    enabled: bool | None = None


class DeveloperTokenRouteRule(BaseModel):
    match_type: str = ""
    method: str | None = None
    path: str = ""


class DeveloperTokenCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    user_id: int
    department_id: int | None = None
    dept_id: str | None = None
    enabled: bool = True
    override_ip_whitelist: bool = False
    ip_whitelist: str | None = ""
    override_rate_limit: bool = False
    rate_limit_per_minute: int | None = None
    route_whitelist: list[DeveloperTokenRouteRule] | None = None
    file_sync_rule: DeveloperTokenFileSyncRule | None = None


class DeveloperTokenUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    user_id: int | None = None
    department_id: int | None = None
    dept_id: str | None = None
    enabled: bool | None = None
    override_ip_whitelist: bool | None = None
    ip_whitelist: str | None = None
    override_rate_limit: bool | None = None
    rate_limit_per_minute: int | None = None
    route_whitelist: list[DeveloperTokenRouteRule] | None = None
    file_sync_rule: DeveloperTokenFileSyncRule | None = None


class DeveloperTokenRead(BaseModel):
    id: int
    tenant_id: int
    tenant_name: str | None = None
    user_id: int
    user_name: str | None = None
    name: str
    token_prefix: str
    enabled: bool
    override_ip_whitelist: bool
    override_rate_limit: bool
    rate_limit_per_minute: int | None = None
    route_rule_count: int = 0
    file_sync_rule: DeveloperTokenFileSyncRule | None = None
    file_sync_target_display: FileSyncTargetDisplay | None = None
    last_used_time: datetime | None = None
    last_used_ip: str | None = None
    created_by: int | None = None
    updated_by: int | None = None
    create_time: datetime | None = None
    update_time: datetime | None = None


class DeveloperTokenDetail(DeveloperTokenRead):
    ip_whitelist: str | None = None
    route_whitelist: list[DeveloperTokenRouteRule] = Field(default_factory=list)


class DeveloperTokenCreateResponse(BaseModel):
    token: DeveloperTokenRead
    plaintext_token: str


class DeveloperTokenSecretResponse(BaseModel):
    id: int
    token_prefix: str
    plaintext_token: str
